src/lib/libcmd/chgrp.c

```diff
@@ -131,7 +131,13 @@ typedef struct Map_s			/* uid/gid map			*/
 	Key_t		to;		/* map to these			*/
 } Map_t;
 
+/*
+ * libast's strgid() and struid() functions will return -1 in response to
+ * not finding an ID for a name on the first call, and -2 for subsequent
+ * calls for the same string.
+ */
 #define NOID		(-1)
+#define STILLNOID	(-2)
```

Comment:
I would dispense with both of these macros and just check for a result < 0 instead.

Reply:
I have pushed a commit that removes these macros. However, when I attempted to test the change after `git clean -fdx` the build completely broke on illumos (and apparently that also broke previous commits). I traced the build failure to commit a874f8c16989079143ceeae52c2ba6ffde1ce3f8. I can confirm after reverting that commit the changes to remove `*NOID` work fine on illumos.

Reply:
Bad news. I guess that'll teach me not to assume that it works on illumos if it works on Solaris. :-/

Reply:
Further testing shows this change is what broke the illumos build:
```diff
--- a/src/cmd/INIT/iffe.sh
+++ b/src/cmd/INIT/iffe.sh
@@ -1369,6 +1369,11 @@ std='#if defined(__STDC__) || defined(__cplusplus) || defined(c_plusplus)
 #define _END_EXTERNS_
 #endif
 #define _NIL_(x)	((x)0)'
+# To ensure the environment tested is the same as that used, add standards
+# compliance macros as probed by libast as soon as they are available.
+if	test -f "${INSTALLROOT}/src/lib/libast/FEATURE/standards"
+then	std=${std}${nl}`cat "${INSTALLROOT}/src/lib/libast/FEATURE/standards"`
+fi
 tst=
 ext="#include <stdio.h>"
 
```

Reply:
The illumos build failure is fixed by removing the `extern swab` feature test from src/ilb/libast/features/sys. The problem is that it checks for this function with void pointers. It doesn't find it because the illumos header uses char pointers instead. And because it doesn't find it, it adds it to ast_sys.h, breaking the build with a 'conflicting declarations' error.

Reply:
The root cause of that is that the _XPG6 standards macro is not getting defined during the tests, so that a legacy definition is exposed. In 580ff6161, I added _XPG6 in features/common. That should be done in features/standards instead! Going to try that now and see if it works.

Reply:
Yup, that was it -- except that illlumos needs _XPG4 instead of _XPG6. We can just define both, it doesn't seem to harm Solaris. So there was nothing wrong with 580ff61 – it exposed breakage, which is a good thing. Thanks for the heads-up.

Reply:
Except now we're getting a new error, introduced in 3082ad87d6c7: `_SC_PAGESIZE` undeclared in the new `astconf_ulong` macro (3082ad87d6c7c2).

Reply:
> Yup, that was it -- except that illlumos needs _XPG4 instead of _XPG6.

I tried doing that on illumos and it just made the build fail even earlier. Removing the `extern swab` is all I need to get ksh building on the current release of OmniOS (r151040), so I've added that to the pull request (I'm not sure what the `_SC_PAGESIZE` error you experienced was caused by as I can't reproduce it, although it could be some form of breakage from defining `_XPG4`).

Reply:
Sorry, but please revert that. It is not a solution.

I've finally figured out (from experimenting and reading their system headers) that to get a non-contradictory build environment on illumos, we need to define `_XPG4` *and* `__EXTENSIONS__` *and* `_XOPEN_SOURCE`. Whereas on Solaris you just define `_XPG6` (or possibly even just `_XPG4`) and everything is fine.

In any case, with all four of those in the correct place (libast/features/standards), it 's now building and running without any further change on both Solaris and Illumos (there are a few regression test failures but I'm pretty sure they were there before). So that's a real fix. Commit coming up.

Reply:
The latest commit on the master branch (https://github.com/ksh93/ksh/commit/2b4c44247a96d0c6d4f9e0f46b39cbef576f856b) is still broken, although the one that was force-pushed out (https://github.com/ksh93/ksh/commit/7a7e027f9936ea63db6a6b17b750f9fdf98a7e73) does work.